roocs_utils/parameter/param_utils.py

```diff
@@ -0,0 +1,190 @@
+from collections.abc import Sequence
+from dateutil import parser as date_parser
+import calendar
+
+from roocs_utils.exceptions import InvalidParameterValue
+from roocs_utils.exceptions import MissingParameterValue
+from roocs_utils.utils.file_utils import FileMapper
+
+
+# Global variables that are generally useful
+month_map = {name.lower(): num for num, name in enumerate(calendar.month_abbr) if num}
+
+time_comp_limits = {
+    "year": None,
+    "month": (1, 12),
+    "day": (1, 40),  # allowing for strange calendars
+    "hour": (0, 23),
+    "minute": (0, 59),
+    "second": (0, 59),
+}
+
+
+# A set of simple parser functions
+def parse_range(x, caller):
+    if isinstance(x, Sequence) and len(x) == 1:
+        x = x[0]
+
+    if x in ("/", None, ""):
+        start = None
+        end = None
+
+    elif isinstance(x, str):
+        if "/" not in x:
+            raise InvalidParameterValue(
+                f"{caller} should be passed in as a range separated by /"
+            )
+
+        # empty string either side of '/' is converted to None
+        start, end = [i.strip() or None for i in x.split("/")]
+
+    elif isinstance(x, Sequence):
+        if len(x) != 2:
+            raise InvalidParameterValue(
+                f"{caller} should be a range. Expected 2 values, " f"received {len(x)}"
+            )
+
+        start, end = x
+
+    else:
+        raise InvalidParameterValue(f"{caller} is not in an accepted format")
+    return start, end
+
+
+def parse_sequence(x, caller):
+
+    if x in (None, ""):
+        sequence = None
+
+    # check str or bytes
+    elif isinstance(x, (str, bytes)):
+        sequence = [i.strip() for i in x.strip().split(",")]
+
+    elif isinstance(x, FileMapper):
+        sequence = [x]
+
+    elif isinstance(x, Sequence):
+        sequence = x
+
+    else:
+        raise InvalidParameterValue(f"{caller} is not in an accepted format")
+
+    return sequence
+
+
+def parse_datetime(dt, default=None):
+    """Parses string to datetime and returns isoformat string for it.
+    If `default` is set, use that in case `dt` is None."""
+    return date_parser.parse(dt, default=default).isoformat()
```

Comment:
@agstephens I've come across an issue with this function, when writing tests for daops. The parsing doesn't work for a 360 day calendar and I get the error `dateutil.parser._parser.ParserError: day is out of range for month: 2007-02-29T12:00:00`. Is this something we were already aware of? I can't find mentions of it in any issues

Reply:
@ellesmith88 please can you look into using replacement for `date_parser.parse()` that _will allow_ any CF datetime? I suspect it exists in `netCDF4`, `cftime`, `nctime` or such packages. (We might already have a solution for this somewhere in our stack :-)

Reply:
see https://github.com/roocs/roocs-utils/pull/88